### PR TITLE
Add FSDP symmetric memory configuration and related tests

### PR DESCRIPTION
**Pull request by crcrpar**
Enable FSDP symmetric-memory communication in TorchTitan’s default trainer path.

This PR wires `parallelism.enable_fsdp_symm_mem` through the regular FSDP2 parallelization paths and calls `FSDPModule.set_symm_mem_for_comm()` on FSDP modules. This covers the default trainer path, not graph trainer or Inductor compiled async-TP symmetric-memory kernels.

### Benchmarks

100-step runs on 4x GB200, default trainer path. Throughput excludes the first 20 warmup steps.

| Model | `torch.compile` | symm_mem off | symm_mem on | Speedup |
|---|---:|---:|---:|---:|
| Llama 3 8B | on | 17,158 tok/s | 17,551 tok/s | 1.023x |
| Llama 3 8B | off | 16,050 tok/s | 16,234 tok/s | 1.011x |
| Qwen3 32B | on | 4,312 tok/s | 4,414 tok/s | 1.024x |
| Qwen3 32B | off | 3,544 tok/s | 3,638 tok/s | 1.027x |
| GPT-OSS 20B | on | 12,839 tok/s | 13,644 tok/s | 1.063x |
| GPT-OSS 20B | off | 9,882 tok/s | 10,220 tok/s | 1.034x |

Used GPT-5.5

### CODE REVIEW

File: tests/integration_tests/features.py [hunk 1]
```diff
@@ -446,6 +446,43 @@ def build_features_test_list() -> list[OverrideDefinitions]:
...
+                [
+                    "--module llama3 --config llama3_debugmodel_float8",
+                    "--parallelism.enable-fsdp-symm-mem",
+                ],
+            ],
+            "FSDP symmetric memory with Float8 all-gather",
+            "fsdp_symm_mem_float8_all_gather",
+            ngpu=2,
+            skip_rocm_test=True,
+            required_packages=("torchao",),
```
Comment (fegin) [inline tests/integration_tests/features.py]: What's the motivation to add this new argument? We should already install torchao for the integration test, iirc.

File: torchtitan/distributed/fsdp.py [hunk 2]
```diff
@@ -25,6 +25,24 @@ def disable_fsdp_gradient_division(model: nn.Module) -> None:
...
+    Enable symmetric-memory communication optimizations for all FSDP modules.
+    """
+    for module in model.modules():
+        if isinstance(module, FSDPModule):
+            if not hasattr(module, "set_symm_mem_for_comm"):
+                raise RuntimeError(
+                    "FSDP symmetric-memory communication was requested, but "
+                    "`FSDPModule.set_symm_mem_for_comm()` is not available in "
+                    "the installed PyTorch build."
+                )
```
Comment (fegin) [inline torchtitan/distributed/fsdp.py]: I don't think this is required, `set_symm_mem_for_comm` has been there for a while. TorchTitan never claims to be BC with older PyTorch version.

**Review by fegin — COMMENTED** [review]
@weifengpy can you review?

File: torchtitan/distributed/fsdp.py [hunk 3]
```diff
@@ -25,6 +25,24 @@ def disable_fsdp_gradient_division(model: nn.Module) -> None:
...
+            # Potential hiccup: TorchTitan also supports float8 FSDP all-gather
+            # via torchao. That path customizes FSDP communication behavior, so
+            # this combination may need extra validation when both are enabled.
+            module.set_symm_mem_for_comm()
```
Comment (tianyu-l) [inline torchtitan/distributed/fsdp.py:35]: Is there not a global flag in FSDP, does it have to happen per FSDPModule?

Comment (weifengpy) [reply torchtitan/distributed/fsdp.py:35]: drafted a PR to support set_symm_mem_for_comm(model, recurse=True): https://github.com/pytorch/pytorch/pull/181659

Comment (weifengpy) [reply torchtitan/distributed/fsdp.py:35]: for @crcrpar, is `set_force_sum_reduction_for_comms` needed here? I remember @kewen mentioned some rough edge for AVG

Comment (weifengpy) [reply torchtitan/distributed/fsdp.py:35]: agree with what fegin pointed out above. other parts looks good to me

File: tests/integration_tests/features.py (hunk 1 above)
Comment (crcrpar) [reply tests/integration_tests/features.py]: I was just paranoid about the torchao availability in CI.

File: torchtitan/distributed/fsdp.py (hunk 3 above)
Comment (crcrpar) [reply torchtitan/distributed/fsdp.py:35]: hmm, I've not tried `FSDPModule.set_force_sum_reduction_for_comms()`

Comment (crcrpar) [reply torchtitan/distributed/fsdp.py:35]: @weifengpy woudl you think I should add `set_force_sum_reduction_for_comms` here, maybe with a config flag?

Comment (weifengpy) [reply torchtitan/distributed/fsdp.py:35]: adding `set_force_sum_reduction_for_comms` should be enough

Comment (weifengpy) [reply torchtitan/distributed/fsdp.py:35]: @crcrpar I didn't see set_force_sum_reduction_for_comms - did you push?

**Review by weifengpy — APPROVED** [review]
looks good on fsdp changes

some nits before landing:
* maybe avoid un-related linter change to make the PR focused?
* I dont see force sum yet. titan use fp32 reduce dtype by default. but would be good to have force sum in case people override default to use bf16/fp16 reduce

File: torchtitan/distributed/fsdp.py (hunk 3 above)
Comment (crcrpar) [reply torchtitan/distributed/fsdp.py:35]: Oops sorry i forgot it. Pushed

**Comment by crcrpar** [comment]
As I can't merge this myself, could any of you land this? maybe @weifengpy?

**Comment by weifengpy** [comment]
fsdp change looks good to me. seems this can only be landed with code-owner review from @fegin @tianyu-l

File: torchtitan/distributed/fsdp.py [hunk 4]
```diff
@@ -25,6 +25,19 @@ def disable_fsdp_gradient_division(model: nn.Module) -> None:
...
 
+def enable_fsdp_symm_mem(model: nn.Module) -> None:
+    """
+    Enable symmetric-memory communication optimizations for all FSDP modules.
+    """
+    for module in model.modules():
+        if isinstance(module, FSDPModule):
+            # Potential hiccup: TorchTitan also supports float8 FSDP all-gather
+            # via torchao. That path customizes FSDP communication behavior, so
+            # this combination may need extra validation when both are enabled.
```
Comment (tianyu-l) [inline torchtitan/distributed/fsdp.py]: float8 FSDP all-gather has been deprecated in torchtitan. Please remove this comment.

File: torchtitan/config/configs.py [hunk 5]
```diff
@@ -119,6 +119,12 @@ class ParallelismConfig:
     - "never" will disable `reshard_after_forward` for all forward passes.
     """
 
+    enable_fsdp_symm_mem: bool = False
+    """
+    Whether to enable FSDP2 symmetric-memory communication optimizations for
+    all FSDP modules after `fully_shard` has been applied.
+    """
```
Comment (tianyu-l) [inline torchtitan/config/configs.py:128]: Does it only work on NVIDIA GPUs? We probably should error out in `ParallelismConfig`'s`post_init`

File: tests/integration_tests/features.py
Comment (tianyu-l) [inline tests/integration_tests/features.py:1]: CI burden is heavy. Maybe just add one more test, or piggyback on an existing test that also skip_rocm, e.g. https://github.com/pytorch/torchtitan/blob/main/tests/integration_tests/features.py#L607

**Comment by tianyu-l** [comment]
Is this real? https://github.com/pytorch/torchtitan/actions/runs/26262664360/job/77303629297?pr=3105

**Comment by crcrpar** [comment]
> Is this real? https://github.com/pytorch/torchtitan/actions/runs/26262664360/job/77303629297?pr=3105

moved the test case to the h100 chunk

**Review by tianyu-l — APPROVED** [review]
